compiler/src/peephole.rs

```diff
@@ -0,0 +1,182 @@
+    }
+
+    fn pop(&mut self) -> (Instruction, InstructionMetadata) {
+        self.buffer.pop().unwrap()
+    }
+
+    fn optimize(&mut self, instruction: Instruction, meta: InstructionMetadata) {
+        macro_rules! lc {
+            ($name:ident {$($field:tt)*}) => {
+                Instruction::LoadConst {
+                    value: bytecode::Constant::$name {$($field)*},
+                }
+            };
+            ($name:ident, $($value:tt)*) => {
+                lc!($name { value: $($value)* })
+            };
+        }
+        macro_rules! emitconst {
+            ([$($metas:expr),*], $($arg:tt)*) => {
+                self.emit(
+                    lc!($($arg)*),
+                    InstructionMetadata::from(vec![$($metas),*]),
+                )
+            };
+        }
+        match instruction {
+            Instruction::BinaryOperation { op, inplace } => {
+                let (rhs, rhs_meta) = self.pop();
+                let (lhs, lhs_meta) = self.pop();
+                macro_rules! op {
+                    ($op:ident) => {
+                        bytecode::BinaryOperator::$op
+                    };
+                }
+                match (op, lhs, rhs) {
+                    (op!(Add), lc!(Integer, lhs), lc!(Integer, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Integer, lhs + rhs)
+                    }
+                    (op!(Subtract), lc!(Integer, lhs), lc!(Integer, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Integer, lhs - rhs)
+                    }
+                    (op!(Add), lc!(Float, lhs), lc!(Float, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Float, lhs + rhs)
+                    }
+                    (op!(Subtract), lc!(Float, lhs), lc!(Float, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Float, lhs - rhs)
+                    }
+                    (op!(Multiply), lc!(Float, lhs), lc!(Float, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Float, lhs * rhs)
+                    }
+                    (op!(Divide), lc!(Float, lhs), lc!(Float, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Float, lhs / rhs)
+                    }
+                    (op!(Power), lc!(Float, lhs), lc!(Float, rhs)) => {
+                        emitconst!([lhs_meta, rhs_meta], Float, lhs.powf(rhs))
+                    }
+                    (op!(Add), lc!(String, mut lhs), lc!(String, rhs)) => {
+                        lhs.push_str(&rhs);
+                        emitconst!([lhs_meta, rhs_meta], String, lhs);
+                    }
+                    (op, lhs, rhs) => {
+                        self.emit(lhs, lhs_meta);
+                        self.emit(rhs, rhs_meta);
+                        self.emit(Instruction::BinaryOperation { op, inplace }, meta);
+                    }
+                }
+            }
+            Instruction::UnpackSequence { size } => {
```

Comment:
This basically starts a second pattern for optimization. Could the two patterns be split into seperate functions? I can see the `optimize` function grow rather large if we continue like this.